### PR TITLE
[bugfix]when replicaDivisionPreference is Weighted and WeightPreference is nil

### PR DESCRIPTION
**Pull request by Poor12**
Signed-off-by: Poor12 <shentiecheng@huawei.com>

**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
Spread by region always failed when replicaDivisionPreference is Weighted and WeightPrefrence is nil.
```
apiVersion: policy.karmada.io/v1alpha1
kind: PropagationPolicy
metadata:
  name: nginx-propagation
spec:
  resourceSelectors:
    - apiVersion: apps/v1
      kind: Deployment
      name: nginx
  placement:
    clusterAffinity:
      clusterNames:
        - member1
        - member2
    replicaScheduling:
      replicaDivisionPreference: Weighted
      replicaSchedulingType: Divided
    spreadConstraints:
    - maxGroups: 1
      minGroups: 1
      spreadByField: region
    - maxGroups: 1
      minGroups: 1
      spreadByField: cluster
```

```
  status:
    conditions:
    - lastTransitionTime: "2022-12-20T03:49:13Z"
      message: Failed to apply all works, see status.aggregatedStatus for details
      reason: FullyAppliedFailed
      status: "False"
      type: FullyApplied
    - lastTransitionTime: "2022-12-20T03:49:14Z"
      message: 'failed to select clusters: the number of feasible region is less than
        spreadConstraint.MinGroups'
      reason: BindingFailedScheduling
      status: "False"
      type: Scheduled
    schedulerObservedGeneration: 9
```
Based on https://github.com/karmada-io/karmada/pull/2451, we skip the `GroupClusterInfo` for region when replicaDivisionPreference is Weighted and WeightPreference is nil. However, we do not change in `shouldIgnoreSpreadConstraint`. So it will come into `SpreadConstraint` logic but the number of feasible region is always 0.  So scheduling failed.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
None

### CODE REVIEW

**Comment by Poor12** [comment]
Need to add e2e for spread-by-region.

File: test/e2e/framework/cluster.go [hunk 1]
```diff
@@ -314,3 +314,49 @@ func LoadRESTClientConfig(kubeconfig string, context string) (*rest.Config, erro
 		loader,
 	).ClientConfig()
 }
+
+// SetClusterRegion set cluster region of E2E
+func SetClusterRegion(c client.Client, clusterName string, regionName string) error {
+	err := wait.PollImmediate(2*time.Second, 10*time.Second, func() (done bool, err error) {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/framework/cluster.go:320]: Directly return will be ok.

File: test/e2e/framework/cluster.go [hunk 2]
```diff
@@ -314,3 +314,49 @@ func LoadRESTClientConfig(kubeconfig string, context string) (*rest.Config, erro
...
+			}
+			return false, err
+		}
+		return true, nil
+	})
+	return err
+}
+
+// DeleteClusterRegion delete cluster region of E2E
+func DeleteClusterRegion(c client.Client, clusterName string) error {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/framework/cluster.go:342]: This func can be replaced with calling `SetClusterRegion(c, xx, "")`

**Comment by Poor12** [comment]
Seems a probabilistic question.

File: test/e2e/spread_by_region_test.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+					}
+
+					if len(targetClusterNames) != updatedRegionGroups {
+						return false, nil
+					}
+					return true, nil
+				}, pollTimeout, pollInterval).Should(gomega.Equal(true))
+			})
+
+			ginkgo.By("update the replica scheduling strategy of propagation policy", func() {
```
Comment (XiShanYongYe-Chang) [inline test/e2e/spread_by_region_test.go:142]: This part of the test looks strange. Can you remove it first?

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks~

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/lgtm
/cc @RainbowMango

**Comment by RainbowMango** [comment]
> Spread by region always failed when replicaDivisionPreference is Weighted and WeightPrefrence is nil.

But, the `WeightPrefrence` is `not` nil in the example?

**Comment by RainbowMango** [comment]
Another question, why `FullyApplied` is present in a situation of scheduling failure?
- lastTransitionTime: "2022-12-19T09:00:03Z"
      message: All works have been successfully applied
      reason: FullyAppliedSuccess
      status: "True"
      type: FullyApplied

**Comment by Poor12** [comment]
> > Spread by region always failed when replicaDivisionPreference is Weighted and WeightPrefrence is nil.
> 
> But, the `WeightPrefrence` is `not` nil in the example?

A mistake in the description. BTW, the `FullyApplied` is not caused by this time of schedule.

**Comment by RainbowMango** [comment]
The YAML in the PR description is not well formatted.

**Review by RainbowMango — APPROVED** [review]
/approved

**Pull request merged**